Update the simplifier to respect and apply the 'use var' user preference.

CyrusNajmabadi opened the pull request:
Note: this PR looks large, but that's primarily due to moving some "use var" functionality down from teh CSharpFeatures layer to the CSharpWorkspace layer. This was done so that the C# simplification service could appropriately simplify code like Dictionary<int, string> d = ... to var d = ... if that matched the user's style preference.

Unfortunately, the classes used for this functionality were not in their own files. They were sibling classes to the "use var/explicit-type' analyzers", and thus show up as huge deletes/adds even though they were barely touched.

CyrusNajmabadi commented:
Tagging @kuhlenh @jinujoseph @dotnet/roslyn-ide

CyrusNajmabadi commented:
Taggign @heejaechang @mavasani as i used IOperation primarily for this.

alrz commented:
would be nice to also recognize
```cs
if (c) M(1);
else M(2);

M(c ? 1 : 2);
```

CyrusNajmabadi commented:
@alrz Not sure how i feel about that one.  If it were to be done, it would be a separate feature IMO.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypeStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -1,186 +0,0 @@
-{
-    internal abstract partial class CSharpTypeStyleDiagnosticAnalyzerBase
-    {
-        internal class State
```
This type was moved down to the CSharpWorkspace layer and renamed to CSharpTypeStyleContext?

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypeStyleDiagnosticAnalyzerBase.cs:
```diff
@@ -1,70 +1,18 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.TypeStyle
 {
-    internal abstract partial class CSharpTypeStyleHelper
```
This type moved down to CSharpWorkspace

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseExplicitTypeDiagnosticAnalyzer.cs:
```diff
@@ -1,157 +1,13 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.TypeStyle
 {
-    internal sealed class CSharpUseExplicitTypeHelper : CSharpTypeStyleHelper
```
This type moved down to CSharpWorkspace

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypeDiagnosticAnalyzer.cs:
```diff
@@ -1,258 +1,13 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Diagnostics.TypeStyle
 {
-    internal sealed class CSharpUseImplicitTypeHelper : CSharpTypeStyleHelper
```
This type moved down to CSharpWorkspace

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs:
```diff
@@ -2272,72 +2280,33 @@ private static bool IsThisOrTypeOrNamespace(MemberAccessExpressionSyntax memberA
                 return false;
             }
 
-            // If the simpleName is the type of the Variable Declaration Syntax belonging to LocalDeclaration, For Statement or Using statement
```
all this code was removed.  We now defer to CSharpTypeStypeHelper to do the logic.  This way we're entirely in sync with the analyzer/refactoring.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/CSharpTypeStyleDiagnosticAnalyzerBase.State.cs:
```diff
@@ -0,0 +1,183 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Utilities
+{
+    internal class CSharpTypeStyleContext
```
This was moved down from CSharpFeatures.  Other than renaming it, it was not changed.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/CSharpTypeStyleHelper.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Utilities
+{
+    internal abstract partial class CSharpTypeStyleHelper
```
This was moved down from CSharpFeatures.  Other than renaming it, it was not changed.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/CSharpUseExplicitTypeHelper.cs:
```diff
@@ -0,0 +1,160 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Utilities
+{
+    internal sealed class CSharpUseExplicitTypeHelper : CSharpTypeStyleHelper
```
This was moved down from CSharpFeatures.  Other than renaming it, it was not changed.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs:
```diff
@@ -0,0 +1,266 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Utilities
+{
+    internal sealed class CSharpUseImplicitTypeHelper : CSharpTypeStyleHelper
```
This was moved down from CSharpFeatures.  Other than renaming it, and making it a singleton, it was not changed.

jcouv commented:
Fixes https://github.com/dotnet/roslyn/issues/23306 ?

Neme12 commented:
> @alrz Not sure how i feel about that one. If it were to be done, it would be a separate feature IMO.

I think he has a good point. This could be generalized to apply to any 2 expressions that are the same except for one subexpression.

CyrusNajmabadi commented:
> I think he has a good point. This could be generalized to apply to any 2 expressions that are the same except for one subexpression.

I see that.  That doesn't mean i think it's a appropriate to fit into this feature :)  I think it's reasonable and palatable to expose the features listed at the top.  If there was something that could do the generalized form, i think it would be another feature altogether that people would have control over.  I am not doing that feature as part of this.  If you'd like, you can open an issue/PR for that if you'd like.

CyrusNajmabadi commented:
It shoudl be done as part of a later PR.  It should be exposed to the user through some other options.  

The implementation is less of a concern for me.  In my opinion it would likely be better handled by some other code, but i don't really care much :)

Neme12 commented on src/Features/Core/Portable/FeaturesResources.resx:
```diff
@@ -1352,4 +1352,16 @@ This version used in: {2}</value>
+    <value>Return statement can be simplified</value>
+  </data>
+  <data name="Simplify_return_statement" xml:space="preserve">
+    <value>Simplify return statement</value>
```
I'm not sure if it's worth to have 2 separate messages for this. I woud imagine something like either "Simplify if statement" or "Use conditional expression". But maybe that's just because I'm thinking of the ways this could be generalized, and having a special message for other cases like "Simplify expression" would not make sense at that point.

Neme12 replied:
I would imagine settings similar to what you did in your parentheses work: "Prefer conditional expression in assignment", "in return statement" and "in other operations". But I don't see why it's needed to have different messages for these.

Neme12 replied:
An example of "other" might be:
```c#
if (condition)
    Console.WriteLine(1 + 2);
else
    Console.WriteLine(1 + 3);
```
to
```c#
Console.WriteLine(1 + (condition ? 2 : 3));
```
I can't come up with a special message for all these cases :/ "Simplify add expression"? In my view these should all have 1 message mentioning either the if statement or conditional expression.

CyrusNajmabadi replied:
Sure, i'll look into seeing if this can be more generalized.

CyrusNajmabadi replied:
Could you start an issue about this other feature?  I would prefer that this PR not be cluttered up by the speculative discussion of what we might do there.

Neme12 replied:
Also, I think it would be good to have a message mentioning "conditional expression" in these regardless of whether we talk about generalizing. "Simplify assignment" doesn't really tell you what the feature is about. If I saw a feature called "simplify assignment", I would probably think it did this:
```c#
a = a + 2;
```
to
```c#
a += 2;
```

Neme12 replied:
As I said, I think either "Simplify if statement" or just "Use conditional expression" would be a better name. With the latter, you'd know exactly what the feature is about just by seeing the message.

Neme12 replied:
"Simplify X" is usually not very descriptive. Especially for a code fix - it might be more acceptable for an analyzer.

CyrusNajmabadi replied:
I tried to generalize this a bit.

CyrusNajmabadi replied:
We use 'Simplify' in several other cases.

Neme12 replied:
@CyrusNajmabadi That's why it would be easy to get in a naming conflict with other features. "Simplify assignment" could apply to anything that somehow manipulates assignment expressions. If we take that name for this, what would you call a feature changing `a = a + 2` to `a += 2`? But I get it that you could still say this is just speculative discussion of what we might do in the future.

CyrusNajmabadi replied:
Yes.  This is all speculative.

CyrusNajmabadi replied:
In general we want to strike a reasonable balance between conveying what type of thing we're going to be doing, without having to be overly verbose and unwieldy with what we say.  We allow the preview "to do the talking" as it were.

So, we've tended to opt for simple high-level verbs to convey ideas.  "Convert to 'a'",  "Simplify 'b'" "Use 'c'" "Generate 'd'", "Make 'e'" etc. etc.

We actually had a bunch of messages that used to try to be more descriptive.  They actually just ended up being way to verbose and unwieldy and cluttered up the UI even *more* than what we ahve today.  In particular we wanted people to know things like "what are we making *and* where is it going?"  So we had things like "Generate field 'Dictionary<int, string> mappings' in class 'MyNamespace.Whatever'".  We pulled way back from that and just settled on letting users know the general idea of what we were going to do, but obviously letting them have hte preview so they could deeply deconstruct what was actually happening.

Neme12 replied:
I can understand that, but I still think "Use conditional expression" is vastly more descriptive without being more verbose.

Neme12 replied:
In my opinion, "Simplify assignment" isn't even on the same level of descriptiveness as "Generate class" - here you know what kind of thing is going to happen (even though it's without more details), but there are likely more ways how assignment could be simplified and I simply don't know what it would mean.

Neme12 replied:
On the other hand, I think it's perfectly fine to have a feature called "Simplify default expression", where it's pretty clear what this will do as there aren't a lot of other options / other forms of a default expression.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Utilities/CSharpUseImplicitTypeHelper.cs:
```diff
@@ -0,0 +1,300 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.Utilities
+{
+    internal sealed class CSharpUseImplicitTypeHelper : CSharpTypeStyleHelper
```
This was moved down from CSharpFeatures. Other than renaming it, and making it a singleton, it was not changed.

Neme12 commented on src/Features/Core/Portable/UseConditionalExpression/UseConditionalExpressionForAssignmentHelpers.cs:
```diff
@@ -0,0 +1,105 @@
+            }
+
+            localDeclarationStatement = parentBlock.Operations[ifIndex - 1] as IVariableDeclarationGroupOperation;
+            if (localDeclarationStatement == null)
```
I'm really surprised by this... this seems like an arbitrary restriction. Why don't we offer to simplify assignment in these cases?
```c#
int local = ...
// many statements
if (condition)
    local = a;
else
    local = b;
```
to
```c#
int local = ...
// many statements
local = condition ? a : b;
```
and in cases where there's no local declaration at all? with parameters? fields? properties?

Neme12 replied:
The fact that *if there happens to be* a declaration right before the if statement, you can later simplify that into one statement, shouldn't disable this code fix in other cases.

CyrusNajmabadi replied:
It isn't intended to.  This was just the code to make sure that we detect the pattern where there is a local-decl right before the if, so we merge it properly.

Neme12 replied:
Thanks

CyrusNajmabadi commented:
@dotnet/roslyn-ide @Neme12 this is ready for review.

CyrusNajmabadi commented on src/EditorFeatures/Test2/Simplification/TypeNameSimplifierTest.vb:
```diff
@@ -1517,7 +1602,7 @@ class Program
         var d = new D();
 
-        foreach (var item in new List&lt;int&gt;()) { }
+        foreach (int item in new List&lt;int&gt;()) { }
```
this is appropriate.  there is no "using System.Collection.Generic", so it was not appropriate for that 'int' to be converted to 'var'.  This changed when i made the simplifier actually use the real "can use var" analysis code.

CyrusNajmabadi commented:
@jcouv  as well, in case you want to review.

CyrusNajmabadi commented:
Note: this PR looks large, but that's primarily due to moving some "use var" functionality down from teh CSharpFeatures layer to the CSharpWorkspace layer.  This was done so that the C# simplification service could appropriately simplify code like ```Dictionary<int, string> d = ...``` to ```var d = ...``` *if* that matched the user's style preference.

Unfortunately, the classes used for this functionality were *not* in their own files.  They were sibling classes to the "use var/explicit-type' analyzers", and thus show up as huge deletes/adds even though they were barely touched.

CyrusNajmabadi commented:
@mavasani @heejaechang can you provide any information on the overhead of using IOperation over doing syntactic checks?  I ask because i'm somewhat concerned about the cost of IOperation causing a lot of churn as each if-statement is checked.  With the SyntaxNode/Semantic approach, we can generally first check syntax to make sure we can bail out quickly, and then only check semantics once the shape looks strongly like something we'd want to check.  But with IOperatoin it's unclear how do split perf in that manner.

Is there any guidance on how to do things well now with IOp?  Thanks!

CyrusNajmabadi commented:
@mavasani Ok, that sounds good!  There's likely going to be at least one iop feature, so it's good that the cost is taken once.

> From a non-analyzer context, say a code fix or a refactoring, it gets tricky as compilation.GetSemanticModel will hand you different semantic model each time, 

This is actually not a problem in the IDE.  The intent of hte workspace API is that if many people are getting the semantic model for a Document at the same time, then they'll get the same instance.  (Though we just fixed a race condition where that wasn't the case).  It's def an important part of our workspace perf story as we tend to have several dozen *independent* features waking up on every edit.  So it's really important that they all be sharing the same semantic model so we're not causing excessive churn inside the compiler.  Glad that that will benefit iop as well.

Thanks!

CyrusNajmabadi commented:
@jinujoseph How would you like to handle this?  Can someone (or two) from roslyn-ide be looped in to help guide this new feature into the product?

CyrusNajmabadi commented:
retest windows_debug_vs-integration_prtest  please

CyrusNajmabadi commented:
@jinujoseph Is there someone who could work with me on this to guide this in?  I believe this is a backlog item for the IDE right now.  Thanks!

jcouv commented:
Yes, I think this fixes https://github.com/dotnet/roslyn/issues/23306

CyrusNajmabadi commented:
@dotnet/roslyn-ide would anyone like to review? 
@DustinCampbell Note: i changed how simplification works a bit now.  I moved the "use var" functionality down from teh CSharpFeatures layer to the CSharpWorkspace layer. This was done so that the C# simplification service could appropriately simplify code like Dictionary<int, string> d = ... to var d = ... if that matched the user's style preference.

This was done because we've had so many features written that always have to duplicate the 'should i use var?' logic in their own way.  Truly, almost every feature just wants to say "i'd like to make a local, and it should match the 'var' preference the user has without me caring about it". Now the system does that.

CyrusNajmabadi commented:
@jcouv can you take a look.  This relates to the work you're doing over in  #26224

CyrusNajmabadi commented on src/Features/CSharp/Portable/TypeStyle/UseExplicitTypeCodeFixProvider.cs:
```diff
@@ -79,7 +79,10 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
+
+                // We're going to be passed through the simplifier.  Tell it to not just convert
+                // this back to var (as that would defeat the purpose of this refactoring entirely).
+                var typeName = typeSymbol.GenerateTypeSyntax(allowVar: false)
```
this same issue will likely affect you @jcouv .  If we have a refactoring to convert 'var' to a proper type, we'll likely want to put the "and do not convert this to var" bit on it so that the simplifier doesn't just undo that work.

CyrusNajmabadi commented:
Tagging @dotnet/roslyn-ide @jcouv @jmarolf .  I've now split this PR out from https://github.com/dotnet/roslyn/pull/26236

This PR is now only the change to move the "use var" logic down to the workspace layer so it can be part of hte simplifier logic.  This is highly beneficial and it means that features do not have to always figure out "should i use 'var' or not".  Instead, as long as they either generate a type with the Simplifier annotation on it (which the core Workspace code generation helpers do), then the Simplifier will convert to 'var' if that's what the user wants and it's legal to do so.

CyrusNajmabadi commented:
Also tagging @DustinCampbell

CyrusNajmabadi commented on src/EditorFeatures/Test2/Simplification/TypeNameSimplifierTest.vb:
```diff
@@ -1467,9 +1468,93 @@ class C1
+{
+    void Main()
+    {
+        {|Simplify:int|} i = 0;
```
these cases were already being tested in this file, but as part of one large test.  I kept hte large test, but also broke out individual cases to make things easier to debug when things go wrong.

CyrusNajmabadi commented on src/Features/CSharp/Portable/CSharpFeaturesResources.Designer.cs:
```diff
@@ -557,7 +557,7 @@ internal class CSharpFeaturesResources {
         }
         
-         /// <summary>
+        /// <summary>
```
no idea why the code was bad before.  but my VS fixed it.

CyrusNajmabadi commented on src/Features/Core/Portable/FeaturesResources.Designer.cs:
```diff
@@ -436,18 +436,16 @@ internal class FeaturesResources {
+            get {
                 return ResourceManager.GetString("Alias_ambiguous_type_0", resourceCulture);
             }
         }
```
someone manually edited this file in teh past.  my VS keeps fixing it, so i'm just going to merge this in since it's a PITA to have to undo this file every time.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CodeStyle/TypeStyle/TypeStyleHelper.cs:
```diff
@@ -306,7 +304,7 @@ public static bool IsPredefinedType(TypeSyntax type)
             }
```
this whole helper can get removed in the future.  features should just ask "type.GenerateTypeSyntax()" and they should get the right behavior.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs:
```diff
@@ -2269,75 +2277,21 @@ private static bool IsThisOrTypeOrNamespace(MemberAccessExpressionSyntax memberA
-            }
-
-            // If the simpleName is the type of the Variable Declaration Syntax belonging to LocalDeclaration, For Statement or Using statement
-            if (simpleName.IsParentKind(SyntaxKind.VariableDeclaration) &&
```
we had duplicated a bunch of "can i use var" logic here.  It was both broken and incomplete. Now, we just defer to the CSharpUseImplicitTypeHelper which contains all the correct logic and is the place we should always be updating.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Simplification/SimplificationOptions.cs:
```diff
@@ -34,7 +34,7 @@ public static class SimplificationOptions
         /// </summary>
-        public static Option<bool> PreferImplicitTypeInLocalDeclaration { get; } = new Option<bool>(nameof(SimplificationOptions), nameof(PreferImplicitTypeInLocalDeclaration), false,
+        public static Option<bool> PreferImplicitTypeInLocalDeclaration { get; } = new Option<bool>(nameof(SimplificationOptions), nameof(PreferImplicitTypeInLocalDeclaration), true,
             storageLocations: new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.PreferImplicitTypeInLocalDeclaration"));
```
this is an option that is not actually exposed to the user in any way.  I'm not sure why we actually have it.

mavasani commented on src/Workspaces/CSharp/Portable/Utilities/CSharpTypeStyleHelper.cs:
```diff
@@ -0,0 +1,125 @@
+        /// things quickly, even if it's going against their stated style.
+        /// </summary>
+        public readonly bool IsStylePreferred;
+        public readonly DiagnosticSeverity Severity;
```
autoprops?

CyrusNajmabadi replied:
This code moved down.  I didn't want to change it.  I also don't see the value in an auto-prop over a readonly field.

mavasani commented on src/Workspaces/CSharp/Portable/Utilities/CSharpTypeStyleHelper.cs:
```diff
@@ -0,0 +1,125 @@
+        /// one exception is the refactoring, which is explicitly there to still let people convert
+        /// things quickly, even if it's going against their stated style.
+        /// </summary>
+        public readonly bool IsStylePreferred;
```
The doc comment is very helpful, but the naming `IsStylePreferred` still seems a bit confusing. I would have assumed that `AnalyzeTypeName(typeSyntax).IsStylePreferred` would tell me if the _current_ type style of `typeSyntax` is preferred, but the current implementation of the flag is opposite. Should this be renamed to `IsConvertedStylePreferred` with current semantics or to `IsExistingTypeStylePreferred` with negated semantics?

CyrusNajmabadi replied:
This code moved down. I didn't want to change it.  :)   I can change the name if you really want though.

mavasani commented on src/Workspaces/CSharp/Portable/Utilities/CSharpUseExplicitTypeHelper.cs:
```diff
@@ -0,0 +1,155 @@
+        {
+        }
+
+        protected override bool IsStylePreferred(
```
Seems like the result `CSharpUseExplicitTypeHelper.IsStylePreferred` is the negation of `CSharpUseImplicitTypeHelper.IsStylePreferred`. Can we share code?

CyrusNajmabadi replied:
All this code moved down.  I did not want to change it.  But i can if you think that's important.  let me know.

mavasani replied:
Ah, I didn't realize that. In that case, I wouldn't do any rename of functionality changes as part of this PR. We can do it later. Thanks.

CyrusNajmabadi commented:
Can this be merged in?  Thanks!

heejaechang commented:
retest windows_debug_vs-integration_prtest please

CyrusNajmabadi commented:
@heejaechang @mavasani All tests passed.  Can this be merged in?  Thanks!

CyrusNajmabadi commented:
Thanks!
